MTV-3931 |  Move VIB validation from populator to provider

amitosw15 opened the pull request:
Change VIB validation from per-populator check to provider-level validation
with a minimum 15-minute interval between checks to improve scalability.

The VIB check is an expensive operation and ESXi has limited capacity for concurrent VIB operations. Previously, each populator pod ran the VIB check independently, often executing many checks in parallel. This caused pod failures and prevented scaling beyond 5 concurrent disk migrations.
Move VIB validation to provider reconciliation with the following behavior:
  1. Check all ESXi hosts during provider reconcile (maximum once per 15 min)
  2. Query VIB version on each host when cache expires
  3. Attempt VIB installation if version does not match desired version
  4. Set warning condition on installation failure (non-blocking)
This eliminates parallel VIB check contention and enables migration of more than 5 disks concurrently.
Resolves: MTV-3931

amitosw15 commented on pkg/lib/vsphere_offload/vib.go:
```diff
@@ -0,0 +1,225 @@
+package vsphere_offload
```
@mmartinv @yaacov 
ping

yaacov replied:
my concern is that this PR moves offload stuff into MTV, AFAIU MTV does is agonistic to offload, we allow it as an external plugin but we should not add offload code inside MTV that is why it's called "offload" e.g. MTV lets other application do the storage movement. any reason for moving offload code inside MTV ?

amitosw15 commented on pkg/controller/plan/vib.go:
```diff
@@ -0,0 +1,94 @@
+)
+
+// validateVIBReadiness validates VIB readiness for migration plans using xcopy volume populators
+func (r *Reconciler) validateVIBReadiness(plan *api.Plan) error {
```
It seems like the common practice and name pattern in validate function, so i copied this behavior.  I agree with you that I can express the meaning better. What you prefer?

amitosw15 committed:
test: add tests and mocks for VIB validation

Add comprehensive test coverage for VIB validation functionality:
- Provider-level VIB validation tests
- Plan-level VIB condition tests
- Mock VIBEnsurer for testing

Resolves: MTV-3931
Signed-off-by: Amit Weinstock <aweinsto@redhat.com>

amitosw15 commented on build/forklift-controller/Containerfile:
```diff
@@ -5,14 +5,16 @@ COPY --chown=1001:0 ./ ./
 ENV GOCACHE=/go-build/cache
-RUN --mount=type=cache,target=${GOCACHE},uid=1001 go build -buildvcs=false -ldflags="-w -s" -o manager github.com/kubev2v/forklift/cmd/forklift-controller
+ARG VIB_VERSION
+RUN --mount=type=cache,target=${GOCACHE},uid=1001 go build -buildvcs=false -ldflags="-w -s -X github.com/kubev2v/forklift/pkg/lib/vsphere_offload.VibVersion=${VIB_VERSION}" -o manager github.com/kubev2v/forklift/cmd/forklift-controller
```
done

amitosw15 replied:
When I changed the Containerfile to use RUN make forklift-controller (to follow the same pattern as vsphere-xcopy-volume-populator and read VIB_VERSION from version.mk), it introduced a dependency chain issue. The forklift-controller target depends on generate, which requires installing controller-gen via go install. However, this installation fails because the Containerfile sets GOFLAGS="-mod=vendor" earlier, preventing external package downloads.

  Solution options:
  1. Install tools before vendor mode (current approach): Install controller-gen without vendor restrictions, then set vendor mode for the actual build
  2. Remove generate dependency: Change the forklift-controller target to only depend on fmt and vet (like vsphere-xcopy-volume-populator does), since generated code is already committed to the repository
  3. Pre-compile outside container: Build the binary locally and copy it into the container (defeats the purpose of using make in the Containerfile)

amitosw15 committed:
Use VIB version from repo instead of CI build args

  Read VIB_VERSION from cmd/vsphere-xcopy-volume-populator/vmkfstools-wrapper/version.mk
  instead of accepting it as an external build argument from CI tools.

Resolves: MTV-3931
Signed-off-by: Amit Weinstock <aweinsto@redhat.com>

amitosw15 commented:
@mnecas @yaacov ping

amitosw15 committed:
Move VIB validation from populator to provider

  Change VIB validation from per-populator check to provider-level validation
  with a minimum 15-minute interval between checks to improve scalability.

  The VIB check is an expensive operation and ESXi has limited capacity for
  concurrent VIB operations. Previously, each populator pod ran the VIB check
  independently, often executing many checks in parallel. This caused pod
  failures and prevented scaling beyond 5 concurrent disk migrations.

  Move VIB validation to provider reconciliation with the following behavior:

  1. Check all ESXi hosts during provider reconcile (maximum once per 15 min)
  2. Query VIB version on each host when cache expires
  3. Attempt VIB installation if version does not match desired version
  4. Set warning condition on installation failure (non-blocking)

  This eliminates parallel VIB check contention and enables migration of more
  than 5 disks concurrently.
Resolves: MTV-3931
Signed-off-by: Amit Weinstock <aweinsto@redhat.com>

amitosw15 committed:
chore: update vendor dependencies for VIB validation

Update govmomi and related vendor dependencies to support VIB validation.

Signed-off-by: Amit Weinstock <aweinsto@redhat.com>

amitosw15 committed:
feat: make VIB validation cache configurable

Make the VIB validation cache duration configurable via the VIB_CACHE_DURATION
environment variable (in minutes, default: 15).

This allows operators to tune the cache interval based on their environment:
- Increase for more stable environments to reduce ESXi load
- Decrease for faster validation during testing/troubleshooting

Also improves cache reliability to ensure validation re-runs correctly
after cache expiration.

Resolves: MTV-3931
Signed-off-by: Amit Weinstock <aweinsto@redhat.com>

amitosw15 commented:
new PRs instead: #6134  #6136
